rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,496 @@
+
+### Passing the port number from VNet to the client
+
+Since VNet creates a single local proxy per app, `alpnproxy.LocalProxy` needs to support dynamically
+changing the SNI per connection.
+
+VNet already uses the [`alpnproxy.LocalProxy.HandleTCPConnector`](https://github.com/gravitational/teleport/blob/3fd79ce5c557cdfa5603b0d691dc8bab199afce2/lib/srv/alpnproxy/local_proxy.go#L249)
+method which was added specifically for VNet. Technically, we could read the port number [in
+`vnet.NetworkStack.handleTCP`](https://github.com/gravitational/teleport/blob/b5df678e22d26776aa87f9dd493b3c443e013fa4/lib/vnet/vnet.go#L377)
+(`req.ID().LocalPort`), but we don't even need to do that. The port number can be read directly
+[from `downstreamConn` inside
+`alpnproxy.LocalProxy.HandleTCPConnector`](https://github.com/gravitational/teleport/blob/3fd79ce5c557cdfa5603b0d691dc8bab199afce2/lib/srv/alpnproxy/local_proxy.go#L267).
+
+Since `alpnproxy.LocalProxy.HandleTCPConnector` is VNet-specific, we can modify it so that the
+config it passes to `dialALPNMaybePing` has a different SNI depending on which local port was used.
+
+### Configuration
+
+The ports are set through the `ports` field. The user is expected to not include a port number in
+the `uri` field. `tctl edit`, `tctl create` and other means of adding apps should be updated to
+prevent the user from including the port number in `uri` if the `ports` field is present. Apps
+without the `ports` field and with no port number in the `uri` field should be allowed to exist in
+order to not introduce a breaking change in behavior – those are technically allowed as of v16.2.
+E.g., Teleport v16.2 lets you define an app with `"tcp://localhost"` as the URI and will only error
+during an actual connection attempt.
+
+In order to keep the implementation simple, the user is allowed to duplicate port numbers and
+ranges. For example, these are all valid `ports` values: `[4080-4090, 4080-4090]`, `[4080-4090,
+4081-4087, 4082]`. The only validation in place is going to be that any given port number and port
+range start and end must be within 1–65535 and that the start of a port range must be less than the
+end, rendering the following not valid: `4090-4080`, `1234-1234`.
+
+#### Limitations
+
+Since a single app spec is supposed to specify available ports, it will not be possible to
+distribute different ports of the same logical Teleport TCP app across different app services. For
+example, had we chosen to go with [collapsing multiple app specs](#multiple-apps-with-the-same-uri-but-different-ports)
+with the same URI but a different port into a single app, it'd have been possible to put each of
+those app specs on a separate app service. This is something customers with an extensive HA setup
+could want Teleport to support.
+
+```
+- name: "foo"
+  uri: "tcp://localhost:3000"
+  labels:
+    env: prod
+```
+
+```
+- name: "foo"
+  uri: "tcp://localhost:3001"
+  labels:
+    env: prod
+```
+
+### Security
```

Comment:
Does an altered SNI require us to make changes to the client TLS config used by the local proxy? Would we have to set InsecureSkipVerify and do custom verification? Since the ServerName we request is unlikely to be in the SANs of the TLS cert presented by the proxy